File: gateway/src/main/java/com/lucasaquiles/micronaut/ProductController.java
package com.lucasaquiles.micronaut;

import io.micronaut.http.annotation.Controller;
import io.micronaut.http.annotation.Get;
import io.reactivex.Flowable;

@Controller("/")
public class ProductController {

    private final ProductFetcher productFetcher;
    private final ProductInventoryFetcher productInventoryFetcher;

    public ProductController(ProductFetcher productFetcher, ProductInventoryFetcher productInventoryFetcher) {
        this.productFetcher = productFetcher;
        this.productInventoryFetcher = productInventoryFetcher;
    }

    @Get("/product")
    Flowable<Product> findAll() {
        return productFetcher.fetchProducts()
                .flatMapMaybe(product -> productInventoryFetcher.inventory(product.getCode())
                .filter(stock -> stock > 0)
                .map(stock -> {
                    product.setStock(stock);
                    return product;
                }));
    }
}

File: product-repository/src/main/kotlin/com/lucasaquiles/micronaut/ProductController.kt
package com.lucasaquiles.micronaut

import io.micronaut.http.HttpResponse
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Get
import io.micronaut.http.annotation.Produces

@Controller("/")
class ProductController {

    @Produces(MediaType.TEXT_PLAIN)
    @Get("/inventory/{code}")
    fun inventory(code: String): HttpResponse<Int> {
        return when (code) {

            "1234" -> HttpResponse.ok(2)
            "4321" -> HttpResponse.ok(3)
            else -> HttpResponse.notFound()
        }
    }
}